### PR TITLE
vscode: save as root fixed

### PR DESCRIPTION
**Pull request by pshirshov**
Just a patch for broken "save as root" functionality taken from here: https://github.com/NixOS/nixpkgs/issues/49643

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137912` at dfddd8a3 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>vscode</li>
    <li>vscode-with-extensions</li>
    <li>vscodium</li>
  </ul>
</details>

File: pkgs/applications/editors/vscode/generic.nix [hunk 1]
```diff
@@ -108,6 +111,30 @@ let
...
+      # as described here https://github.com/NixOS/nixpkgs/blob/fc553c0bc5411478e2448a707f74369ae9351e96/pkgs/tools/misc/etcher/default.nix#L49y
+      sed -i "
+        s|/usr/bin/pkexec|/run/wrappers/bin/pkexec|
+        s|/bin/bash|${bash}/bin/bash|
+      " "$unpacked/sudo-prompt/index.js"
+      # delete original .asar file, as the new unpacked is now replacing it
+      rm -rf "$packed"
+      # PATCHING GLOBAL SEARCH
+      # ----------------------
+      chmod +x resources/app/node_modules/vscode-ripgrep/bin/rg
```
Comment (turion) [inline pkgs/applications/editors/vscode/generic.nix:135]: That sounds like it should be part of a different PR?

File: pkgs/applications/editors/vscode/generic.nix [hunk 2]
```diff
@@ -108,6 +111,30 @@ let
...
+      # where to find the node.js librar .asar was described here https://github.com/NixOS/nixpkgs/issues/76526#issuecomment-569131432
+      packed="resources/app/node_modules.asar"
+      # we unpack directly into the same name without .asar ending,
+      # which is adapted from hacking-skype-tutorial https://www.codepicky.com/hacking-electron-restyle-skype/
+      unpacked="resources/app/node_modules"
+
+      ${nodePackages.asar}/bin/asar extract "$packed" "$unpacked"
+      # we change paths to pkexec and bash
+      # as described here https://github.com/NixOS/nixpkgs/blob/fc553c0bc5411478e2448a707f74369ae9351e96/pkgs/tools/misc/etcher/default.nix#L49y
+      sed -i "
```
Comment (turion) [inline pkgs/applications/editors/vscode/generic.nix:127]: Couldn't you use `patchShebangs`?

File: pkgs/applications/editors/vscode/generic.nix [hunk 3]
```diff
@@ -108,6 +111,30 @@ let
       )
     '';
 
+    # See https://github.com/NixOS/nixpkgs/issues/49643#issuecomment-873853897
+    postPatch = ''
+      # PATCHING RUN AS SUDO
+      # --------------------
+      # where to find the node.js librar .asar was described here https://github.com/NixOS/nixpkgs/issues/76526#issuecomment-569131432
```
Comment (turion) [inline pkgs/applications/editors/vscode/generic.nix:118]:
```suggestion
      # where to find the node.js library .asar was described here https://github.com/NixOS/nixpkgs/issues/76526#issuecomment-569131432
```

**Review by turion — CHANGES_REQUESTED** [review]
Some small comments. Thanks for moving this forward!

File: pkgs/applications/editors/vscode/generic.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/applications/editors/vscode/generic.nix:135]: Or commit

File: pkgs/applications/editors/vscode/generic.nix (hunk 2 above)
Comment (pshirshov) [reply pkgs/applications/editors/vscode/generic.nix:127]: If I only know what is it and how to do that.

In fact I've just PRed a very useful patch I found around as is.

File: pkgs/applications/editors/vscode/generic.nix (hunk 1 above)
Comment (pshirshov) [reply pkgs/applications/editors/vscode/generic.nix:135]: Well, this patch fixes them both, I'm not sure why is it necessary to split.

Comment (turion) [reply pkgs/applications/editors/vscode/generic.nix:135]: Because they fix different issues. If we have to revert one of them, for example, we don't want to revert the other one. Or if we want to bisect to search for a fix or a bug, it is important that they are separate.

File: pkgs/applications/editors/vscode/generic.nix (hunk 2 above)
Comment (turion) [reply pkgs/applications/editors/vscode/generic.nix:127]: > In fact I've just PRed a very useful patch I found around as is.

It's great that you created a PR so the issue will be fixed for everyone (and not just those that adapt the patch into their overlay). If it's added to nixpkgs though, the code quality has to be high, so I'm trying to help you achieve a high quality of the PR.

> If I only know what is it and how to do that.

`patchShebangs` replaces paths like `/usr/bin/something` with the appropriate nix store path like `${pkgs.something}/bin/something`. It's used in a lot of places in nixpkgs [as one can see in the search](https://github.com/NixOS/nixpkgs/search?q=patchShebangs). See there for examples.

I think that it might be able to fix this in a more standard way.

Comment (Artturin) [reply pkgs/applications/editors/vscode/generic.nix:127]: patchshebangs only replaces the shebang !#

Comment (Synthetica9) [reply pkgs/applications/editors/vscode/generic.nix:127]: Yeah, in this case you want substituteInPlace

**Comment by pshirshov** [comment]
Well, I've splitted this into two commits. Hope it's good enough now.

**Comment by Artturin** [comment]
fyi when nixpkgs-fmt does lots of formatting its usually done in separate commit so that git blame still works

**Comment by pshirshov** [comment]
I've already removed irrelevant changes.

**Comment by pshirshov** [comment]
So, is it good enough to be merged now?

**Comment by pshirshov** [comment]
Also, inb4, I've tested the last version of the patch, it works.

File: pkgs/applications/editors/vscode/generic.nix [hunk 4]
```diff
@@ -108,6 +111,18 @@ let
...
 
+    # See https://github.com/NixOS/nixpkgs/issues/49643#issuecomment-873853897
+    postPatch = ''
+      # this is a fix for "save as root" functionality
+      packed="resources/app/node_modules.asar"
+      unpacked="resources/app/node_modules"
+      ${nodePackages.asar}/bin/asar extract "$packed" "$unpacked"
+      substituteInPlace $unpacked/sudo-prompt/index.js \
+        --replace "/usr/bin/pkexec" "/run/wrappers/bin/pkexec" \
+        --replace "/bin/bash" "${bash}/bin/bash"
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/generic.nix:122]: Do we need readline support? If so please use bashInteractive.

File: pkgs/applications/editors/vscode/generic.nix [hunk 5]
```diff
@@ -108,6 +111,18 @@ let
...
+      rm -rf "$packed"
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/generic.nix:123]: If this does not exist we should error, please remove the -f.

File: pkgs/applications/editors/vscode/generic.nix [hunk 6]
```diff
@@ -108,6 +111,18 @@ let
       )
     '';
 
+    # See https://github.com/NixOS/nixpkgs/issues/49643#issuecomment-873853897
+    postPatch = ''
+      # this is a fix for "save as root" functionality
+      packed="resources/app/node_modules.asar"
+      unpacked="resources/app/node_modules"
+      ${nodePackages.asar}/bin/asar extract "$packed" "$unpacked"
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/generic.nix:119]: Why not use nativeBuildInputs?

Comment (pshirshov) [reply pkgs/applications/editors/vscode/generic.nix:119]: I have no idea how to/what is it but you are very welcome to propose a change. I would be happy to learn something new about nix.

File: pkgs/applications/editors/vscode/generic.nix (hunk 5 above)
Comment (pshirshov) [reply pkgs/applications/editors/vscode/generic.nix:123]: Ok, will do.

File: pkgs/applications/editors/vscode/generic.nix (hunk 4 above)
Comment (pshirshov) [reply pkgs/applications/editors/vscode/generic.nix:122]: Could you suggest the necessary change? I have, ehm, zero Nix experience.

Comment (Artturin) [reply pkgs/applications/editors/vscode/generic.nix:122]: Pshirshov said he tested it so I assume we don't

File: pkgs/applications/editors/vscode/generic.nix (hunk 6 above)
Comment (Artturin) [reply pkgs/applications/editors/vscode/generic.nix:119]: > I have no idea how to/what is it but you are very welcome to propose a change. I would be happy to learn something new about nix.

If nodePackages.asar is put in nativeBuildInputs then the asar command would be added to the build path so the command would be `asar extract` 

Imo it's fine as it is, no change needed.

Comment (SuperSandro2000) [reply pkgs/applications/editors/vscode/generic.nix:119]: Not using hardcored paths in build environments is preferred. Hardcore paths are more likely to break and slip by treewide changes related to buildInputs. It is totally fine to use absolute paths in patching/build scripts etc.
